crates/krun-server/src/bin/krun-server.rs

```diff
@@ -1,32 +1,96 @@
 use std::net::TcpListener;
-use std::os::fd::AsRawFd;
-use std::panic;
+use std::os::unix::process::ExitStatusExt as _;
 use std::process::Command;
+use std::sync::mpsc::{channel, TryRecvError};
+use std::sync::{Arc, Condvar, Mutex};
 
-use anyhow::{Context, Result};
+use anyhow::Result;
 use krun_server::cli_options::options;
 use krun_server::server::start_server;
-use nix::sys::socket::{shutdown, Shutdown};
+use log::debug;
 
 fn main() -> Result<()> {
     env_logger::init();
 
     let options = options().run();
 
     let listener = TcpListener::bind(format!("0.0.0.0:{}", options.server_port))?;
-    let listener_fd = listener.as_raw_fd();
+    let (child_tx, child_rx) = channel();
+    let idle_pair = Arc::new((Mutex::new(true), Condvar::new()));
 
-    let server_thread = start_server(listener);
+    start_server(listener, child_tx, Arc::clone(&idle_pair));
 
-    Command::new(&options.command)
+    match Command::new(&options.command)
         .args(options.command_args)
         .status()
-        .with_context(|| format!("Failed to execute command {:?}", options.command))?;
-
-    shutdown(listener_fd, Shutdown::Both)?;
-    if let Err(err) = server_thread.join() {
-        panic::resume_unwind(err);
+    {
+        Ok(status) => {
+            if !status.success() {
+                if let Some(code) = status.code() {
+                    eprintln!(
+                        "{:?} process exited with status code: {code}",
+                        options.command
+                    );
+                } else {
+                    eprintln!(
+                        "{:?} process terminated by signal: {}",
+                        options.command,
+                        status
+                            .signal()
+                            .expect("either one of status code or signal should be set")
+                    );
+                }
+            }
+        },
+        Err(err) => {
+            eprintln!(
+                "Failed to execute {:?} as child process: {err}",
+                options.command
+            );
+        },
     }
 
-    Ok(())
+    println!("Waiting for other commands launched through this krun server to exit...");
+    println!("Press Ctrl+C to force quit");
```

Comment:
Ideally, this should only be printed if we're really going to wait on some child.

Reply:
We could print the message the first time we're going to `wait` for a child. WDYT?

Reply:
Done.